WIP: Cluster support

alisaifee opened the pull request:
# Description

Initial attempt at adding cluster support

## TODO 
- [ ] Fix cluster specific tests in `tests/cluster`
- [ ] Pubsub
- [ ] ...

alisaifee committed:
Allow gather to return exceptions

This mirrors the asyncio.gather functionality to be able to
still wait for all tasks even if some fail.

alisaifee committed:
Add utility AsyncQueue

This can be used a a simple replacement for an asyncio.FifoQueue
which works for both asyncio & trio

alisaifee committed:
Refactor cluster client/connection pool for anyio

- Update cluster client/connection pool to use anyio semantics
  and expose a similar context manager use case as the single
  redis client
- Replace use of asyncio.Queue with coredis._async_utils.AsyncQueue
  (Which is a light wrapper around a memory object stream)
- Update Cluster pipeline implementation to be consistent with single
 client pipeline

Graeme22 committed:
uncouple pool & cxn, restructure pool

Graeme22 committed:
fix corrupted pubsub issue

Graeme22 commented on coredis/_utils.py:
```diff
@@ -500,22 +503,48 @@ async def gather(
 
 
-async def gather(*awaitables: Awaitable[Any]) -> tuple[Any, ...]:
-    if len(awaitables) == 1:
```
Can we keep this optimization? I think it's worth it to not have to create a task group in the first place

alisaifee replied:
Addressed in 12d2af7397a9bbe1243f14be036d65e3ec5ef760

Graeme22 commented on coredis/_utils.py:
```diff
@@ -500,22 +503,48 @@ async def gather(
+        nonlocal first_exc
+        try:
+            results[i] = await awaitable
+        except BaseException as exc:
```
Since we're not re-raising here it's probably best to catch Exception instead of BaseException to prevent unexpected behavior (CancelledError if caught should always be re-raised)

alisaifee replied:
Simplified in 12d2af7397a9bbe1243f14be036d65e3ec5ef760

Graeme22 commented on coredis/_utils.py:
```diff
@@ -500,22 +503,48 @@ async def gather(
+            results[i] = await awaitable
+        except BaseException as exc:
+            results[i] = exc
+            if first_exc is None:
```
This might be cleaner to just have two different runner wrappers, one that catches exceptions and one that doesn't

alisaifee replied:
Instead of that I've simplified the runner here: 12d2af7397a9bbe1243f14be036d65e3ec5ef760

Graeme22 commented on coredis/client/cluster.py:
```diff
@@ -639,14 +640,16 @@ def from_url(
+        async with self.connection_pool:
+            await self._populate_module_versions()
+            if self.cache:
+                await self.cache.initialize(self)
```
To be consistent the cache should probably use a context manager as well

Graeme22 commented on coredis/client/cluster.py:
```diff
@@ -1164,7 +1163,7 @@ def sharded_pubsub(
-    async def pipeline(
+    def pipeline(
         self,
         transaction: bool = False,
```
In the basic client this defaults to True so one of them should be changed to be consistent

alisaifee replied:
yeah this is a bit annoying... transactions only work in cluster if all the commands in the pipeline route to the same node. To be honest the whole pipelining approach in cluster mode is riddled with foot guns!

Graeme22 commented on tests/cluster/test_pipeline.py:
```diff
@@ -322,82 +323,17 @@ async def test_multi_node_pipeline_partially_correct(self, client):
-        assert results[0] == (True,)
-        assert int(await client.get("c{fubar}")) > 3
-
-    async def test_transaction_callable_access_other_node(self, client, cloner):
```
Just curious what's the reason for removing these?

alisaifee replied:
The `transaction` helper was removed in 8c6dfa1a0553332b0d6ea7e4192786ad4eb9e9c3

Graeme22 replied:
Oh right!

alisaifee committed:
Simplify implementation of gather

- Add back optimization for single awaitable
- Simply raise from runner if an exception is encountered when
  return_exceptions is False instead of using cancel scope

Graeme22 committed:
pubsub reconnection logic

alisaifee committed:
Ensure per request encoding is respected

alisaifee committed:
Ensure noreply commands do not enqueue requests to be parsed

Graeme22 commented:
Hi @alisaifee why did you close this?

alisaifee commented:
> Hi @alisaifee why did you close this?

The PR was just to share the initial connection pool approach for cluster in an easy to view format.
